fs.go

```diff
@@ -208,3 +210,99 @@ func CopyFile(src string, dest string) error {
 
 	return os.Chmod(dest, srcinfo.Mode())
 }
+
+// HasPrefix will determine if "path" starts with "prefix" from the
+// point of view of a filesystem.
+//
+// Unlike strings.HasPrefix, this functions is path-aware, meaning that
+// it knows that two directories /foo and /foobar and not the same
+// thing, and therefore HasPrefix("/foobar", "/foo") will return false.
+func HasPrefix(path, prefix string) bool {
+	if filepath.VolumeName(path) != filepath.VolumeName(prefix) {
+		return false
+	}
+
+	var dn string
+
+	if isDir, err := IsDir(path); err != nil {
+		return false
+	} else if isDir {
+		dn = path
+	} else {
+		dn = filepath.Dir(path)
+	}
+
+	dn = strings.TrimSuffix(dn, string(os.PathSeparator))
+	prefix = strings.TrimSuffix(prefix, string(os.PathSeparator))
+
+	dirs := strings.Split(dn, string(os.PathSeparator))[1:]
+	prefixes := strings.Split(prefix, string(os.PathSeparator))[1:]
+
+	if len(prefixes) > len(dirs) {
+		return false
+	}
+
+	var d, p string
+
+	for i := range prefixes {
+		// need to test each component of the path for
+		// case-sensitiveness because on Unix we could have
+		// something like ext4 filesystem mounted on FAT
+		// mountpoint, mounted on ext4 filesystem, i.e. the
+		// problematic filesystem is not the last one.
+		if isCaseSensitiveFilesystem(filepath.Join(d, dirs[i])) {
+			d = filepath.Join(d, dirs[i])
+			p = filepath.Join(p, prefixes[i])
+		} else {
+			d = filepath.Join(d, strings.ToLower(dirs[i]))
+			p = filepath.Join(p, strings.ToLower(prefixes[i]))
+		}
+
+		if p != d {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isCaseSensitiveFilesystem(dir string) bool {
+	// build a path where the last element has all the runes
+	// case-flipped, so that if we get /foo/BaR as input we will
+	// test for /foo/bAr. This takes care of inputs like /foo/bar
+	// and /foo/BAR.
+	alt := filepath.Join(filepath.Dir(dir), strings.Map(func(r rune) rune {
+		if unicode.IsLower(r) {
+			return unicode.ToUpper(r)
+		} else {
```

Comment:
I'm far from a unicode expert (and the little dive I just did was terrifying), but I am sure there are unicode code points that are caseless/uncased: [for example](https://play.golang.org/p/83lhJ_tRMJ).

For these purposes, I think the best thing to do is probably to explicitly check `unicode.IsUpper()` here. If both are false, simply return the original input rune. (No need for a third branch, `else` with a return is pointless).